### PR TITLE
[Query] Add common library to support MQTT-hybrid protocol @open sesame 09/29 10:29

### PR DESCRIPTION
**Pull request by again4you**
This PR includes a common library to support the MQTT-hybrid protocol. This library support both raw and TCP message for edge servers' information.  It supports retained messages so the client always gets the latest information of the edge server when just starting to subscribe to the target topic.

### Related patch
* [query] Add common library to support MQTT-hybrid protocol
* [test/query] Add test case to publish & subscribe specific topic message

### Message Data structure
```cpp
typedef struct _msg_data {
  msg_type_t type;                          /**< 4 byte: message type */
  union {
    /* for raw data */
    struct {
      uint8_t payload[MAX_MSG_SIZE];        /* 1020 byte: payload */
    };

    /* for TCP data */
    struct {
      uint8_t srv_host[HOST_NAME_MAX];      /**< 64 byte: server host name */
      uint16_t port;                        /**< 2 byte: port number */
      tcp_server_state_t state;             /**< 4 byte: tcp server state */
      uint8_t custom[MAX_TCP_CUSTOM_SIZE];  /**< 950 byte: custom data (optional) */
    } __attribute__((aligned(1), packed));;
  };
} msg_data;
```

### Usage
* Publish TCP info of edge server
```cpp
char msg[] = "test custom message";
query_h handle;
int ret = 0;

// Default MQTT broker / port is used and state change callback is not registered.
ret = query_open_connection(&handle, NULL, NULL, NULL, NULL);
...

// Topic: nns_query/tcp/info
// edge server host: localhost
// edge server port: 12345
// edge server status: TCP_BUSY
// custom message (optional): "test custom message"
ret = query_publish_tcp_data (handle, "nns_query/tcp/info",
      "localhost", 12345, TCP_BUSY, msg, strlen(msg), true);
...

ret = query_close_connection(handle);
```

* Subscribe TCP info of edge server
```cpp
static void
tcp_msg_received_cb (const char *topic_name,
    msg_data *msg, int msg_len, void *user_data)
{
  char buf[MAX_TCP_CUSTOM_SIZE] = { 0 };
  int custom_size = msg_len - TCP_BASIC_SIZE;
  if (custom_size > 0 && msg->custom) {
    memcpy (buf, msg->custom, custom_size);
    buf[custom_size] = '\0';
  }

  printf ("Received Topic: %s (Size: %d)\n", topic_name, msg_len);
  printf (" - type: %d\n", msg->type);
  printf (" - server hostname: %s\n", msg->srv_host);
  printf (" - port: %u\n", msg->port);
  printf (" - state: %d\n", msg->state);
  printf (" - custom: %s\n", buf);
}

ret = query_open_connection(&handle, NULL, NULL, NULL, NULL);

// topic: nns_query/tcp/info
// callback function: tcp_msg_received_cb
// user_data: NULL
ret = query_subscribe_topic (handle, "nns_query/tcp/info", tcp_msg_received_cb, NULL);

... // main loop is running

ret = query_close_connection(handle);
```

### CODE REVIEW

**Comment by gichan-jang** [comment]
Q. If the query server sends tcp information to the broker, how long is it kept?

File: tests/test_nnsquery.c [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+      printf ("Error: query_close_connection() ret: %d\n", ret);
+      return -1;
+  }
+  sleep(2);
+
+  return 0;
+}
+
+int
+tc_subscribe_tcp_message ()\
```
Comment (gichan-jang) [inline tests/test_nnsquery.c:165]: Unnecessary character
```suggestion
tc_subscribe_tcp_message ()
```

File: src/libquery/nnsquery.c [hunk 2]
```diff
@@ -0,0 +1,503 @@
+/* SPDX-License-Identifier: LGPL-2.1-only */
```
Comment (myungjoo) [inline src/libquery/nnsquery.c:1]: Let's use Apache 2.0

File: src/libquery/nnsquery.h [hunk 3]
```diff
@@ -0,0 +1,149 @@
+/* SPDX-License-Identifier: LGPL-2.1-only */
```
Comment (myungjoo) [inline src/libquery/nnsquery.h:1]: LGPL ---> Apache

File: src/libsensor/edge_sensor.c [hunk 4]
```diff
@@ -0,0 +1,516 @@
+/* SPDX-License-Identifier: LGPL-2.1-only */
```
Comment (myungjoo) [inline src/libsensor/edge_sensor.c:1]: LGPL --> Apache

File: src/libsensor/edge_sensor.h [hunk 5]
```diff
@@ -0,0 +1,75 @@
+/* SPDX-License-Identifier: LGPL-2.1-only */
```
Comment (myungjoo) [inline src/libsensor/edge_sensor.h:1]: LGPL-->Apache

File: tests/test_edge_sensor.c [hunk 6]
```diff
@@ -0,0 +1,147 @@
+
```
Comment (myungjoo) [inline tests/test_edge_sensor.c:1]: Add Apache SPDX

File: tests/test_nnsquery.c [hunk 7]
```diff
@@ -0,0 +1,230 @@
+/**
```
Comment (myungjoo) [inline tests/test_nnsquery.c:1]: Add Apache SPDX

**Comment by myungjoo** [comment]
Let's locate external-headers somewhere independent. (e.g., /include )
And keep private-headers w/ source files.

File: tests/test_nnsquery.c [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+  ret = query_open_connection (&handle, NULL, NULL, NULL, NULL);
+  if (ret != 0) {
+      printf ("Error: query_open_connection() ret: %d\n", ret);
+      return -1;
+  }
+  sleep(2);
+
+  ret = query_publish_raw_data (handle, "nns_query/tcp/info", msg, strlen(msg), true);
+  if (ret != 0) {
+    printf ("Error: query_close_connection() ret: %d\n", ret);
```
Comment (gichan-jang) [inline tests/test_nnsquery.c:85]: Please fix the error message here and other lines.

File: tests/test_nnsquery.c [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+#include <string.h>
+#include "nnsquery.h"
+
+static void
+raw_msg_received_cb (const char *topic_name,
+    msg_data *msg, int msg_len, void *user_data)
+{
+  char *buf;
+  int size = msg_len - sizeof(msg->type);
+  buf = (char *)malloc(size + 1);
```
Comment (gichan-jang) [inline tests/test_nnsquery.c:18]: Need to release the resource.

**Comment by jaeyun-jung** [comment]
Now we implemented nns-edge interfaces, close this PR.

**Pull request closed without merging**